README.md
# Simple Tip

[![codecov](https://codecov.io/gh/evan361425/flutter-simple-tip/branch/master/graph/badge.svg?token=2Lx6Ozhzjw)](https://codecov.io/gh/evan361425/flutter-simple-tip)
[![Codacy Badge](https://app.codacy.com/project/badge/Grade/29d3892a0deb4ea79a6def655f5adae1)](https://www.codacy.com/gh/evan361425/flutter-simple-tip/dashboard?utm_source=github.com&utm_medium=referral&utm_content=evan361425/flutter-simple-tip&utm_campaign=Badge_Grade)
[![Pub Version](https://img.shields.io/pub/v/simple_tip)](https://pub.dev/packages/simple_tip)

A widget for showing tips to it's child that automatically setup position.

> This package is separated from my project [POS-System](https://github.com/evan361425/flutter-pos-system).

<img src="images/example.gif" alt="SimpleTip DEMO" width="300"/>

## How to use

### All in one screen

Wrap your widget with [`SimpleTip`](#simpletip), and give some tips!

```dart
Widget build(context) {
  return SimpleTip(
    message: 'This is my tip!',
    child: Text('Hello world'),
  );
}
```

### In order

Wrap your widget with [`OrderedTip`](#orderedtip) and give `grouper` and `id` for each tip.

```dart
Widget build(context) {
  final grouper = GlobalKey<TipGrouperState>();
  final observer = RouteObserver<ModalRoute<void>>();

  return TipGrouper(
    key: grouper,
    id: 'group',
    candidateLength: 2,
    routeObserver: observer,
    child: Scaffold(
      body: Column(children: [
        OrderedTip(
          order: 1,
          version: 1,
          id: 't-1',
          message: 'Tip content 1',
          grouper: grouper,
          child: const Card(
            child: const Text('First tip'),
          ),
        ),
        OrderedTip(
          order: 2,
          version: 1,
          id: 't-2',
          message: 'Tip content 2',
          grouper: grouper,
          child: const Text('Second tip'),
        ),
      ]),
    ),
  );
}
```

> You SHOULD setup [`defaultStateManager`](#state-manager), otherwise it only works in-memory.
> Which means when you restart your app, tips will show up again!

## Configuration

Two main widget you can use, [`SimpleTip`](#simpletip) and [`OrderedTip`](#orderedtip).

> You can follow Dart provided [API document](https://pub.dev/documentation/simple_tip/latest/simple_tip/simple_tip-library.html)

### SimpleTip

- `title`, `String?`, `null`

  - The title to display in the tip.

- `message`, `String?`, `null`

  - The message to display in the tip. **MUST** set if `content` not set.
  - This value will also use for semantics.

- `content`, `Widget?`, `null`
  - Content of the tip.
  - This will "win" when `message` and `content` both set.
  - Example:

```dart
final content = Text('hi');
```

- `boxConstraints`, `BoxConstraints?`, `BoxConstraints(minHeight: 24.0)`

  - Tip's content constraints

- `decoration`, `Decoration?`, see below
  - Tip's container decoration
  - Default:

```dart
ShapeDecoration(
  color: (isDark ? Colors.white : Colors.grey[700]!).withOpacity(0.9),
  shape: TipShapeBorder(arrowArc: 0.1, target: target),
)
```

- `textStyle`, `TextStyle?`, see below
  - Tip's text default style
  - Default:

```dart
Theme.of(context).textTheme.bodyText1!.copyWith(
  color: isDark ? Colors.black : Colors.white,
)
```

- `onClosed`, `VoidCallback?`, `null`
  - A callback after the start of closing tip.
  - This will be needed if you handle tip by filesystem, eg: SharedPreferences, hive
  - Example:

```dart
final isDisabled = instance.read('myTip') == 1;
final onClosed = () {
  instance.write('myTip', 1);
}
```

- `padding`, `EdgeInsets`, `EdgeInsets.all(8.0)`

  - The amount of space by which to inset the tip's content.

- `margin`, `EdgeInsets`, `EdgeInsets.symmetric(horizontal: 16.0)`

  - The empty space that surrounds the tip.
  - Defines the tip's outer `Container.margin`. By default, a long tip will span the width of its window. If long enough, a tip might also span the window's height. This property allows one to define how much space the tip must be inset from the edges of their display window.

- `verticalOffset`, `double`, `24.0`

  - The vertical gap between the widget and the displayed tip.
  - When `preferBelow` is set to true and tips have sufficient space to display themselves, this property defines how much vertical space tips will position themselves under their corresponding widgets. Otherwise, tips will position themselves above their corresponding widgets with the given offset.

- `closerText`, `String`, `OK`

  - Text of button to close tip.

- `waitDuration`, `Duration`, `Duration.zero`

  - The length of time that a tip will wait for showing.
  - Defaults to 0 milliseconds (tips are shown immediately after created).

- `excludeFromSemantics`, `bool`, `false`

  - Whether the tip's `SimpleTip.message` should be excluded from the semantics tree.
  - A tip will add a `Semantics` label that is set to `SimpleTip.message`. Set this property to true if the app is going to provide its own custom semantics label.

- `isDisabled`, `bool`, `false`

  - Disable tip.
  - Wrap `SimpleTip` with `StatefulWidget` and dynamically set this value.

- `withBackdrop`, `bool`, `false`

  - Whether use backdrop

- `preferBelow`, `bool`, `true`

  - Whether the tips defaults to being displayed below the widget.
  - Defaults to true. If there is insufficient space to display the tip in the preferred direction, the tip will be displayed in the opposite direction.

- `child`, `Widget`, **required**
  - The widget below this widget in the tree.

### OrderedTip

> **IMPORTANT** you should set up TipGrouper static property `defaultStateManager` to "remember" that user had read this tip.

- `id`, `String`, **required**

  - ID of this tip.
  - It should be unique in the same group

- `grouper`, `GlobalKey<TipGrouperState>`, **required**

  - State of grouper that contains many [OrderedTip]
  - If one screen have multiple groups, it should show many tips in one screen.

- `version`, `int`, `0`
  - The version it should be.
  - `SimpleTip.isDisabled` will be `false` if version is not equal to given version from `static getVersion`.
  - Implement detail:

```dart
if (TipOrdered.getVersion(groupId, id) != version) {
  enabledTip = id;
  break; // break the loop
}
```

- `order`, `int`, `0`
  - The order to show the tip, lower order higher priority.
  - Implement:

```dart
tipList.sort((a, b) => a.order.compareTo(b.order))
```

- `title`, `String?`, same as `SimpleTip.title`
- `message`, `String?`, same as `SimpleTip.message`
- `content`, `Widget?`, same as `SimpleTip.content`

#### State Manager

There are two method you need to override:

- `shouldShow`, `int Function(String groupId, TipItem item)`, see below
  - Get `OrderedTip.version` from your filesystem, eg: SharedPreferences, hive
  - Default using in-memory data to get version:

```dart
bool shouldShow(String groupId, TipItem item) {
  final lastVersion = _records['$groupId.${item.id}'];
  return lastVersion == null ? true : lastVersion < item.version;
}
```

- `tipRead`, `Future<void> Function(String groupId, TipItem item)`, see below
  - Set `OrderedTip.version` after user manually close it
  - Default using in-memory data to record version:

```dart
Future<void> tipRead(String groupId, TipItem item) async {
  _records['$groupId.${item.id}'] = item.version;
}
```

Example of using [shared_preferences](https://pub.dev/packages/shared_preferences):

```dart
void initialize() async {
  final service = await SharedPreferences.getInstance();
  OrderedTip.stateManager = PrefStateManager(service);
}

class PrefStateManager extends StateManager {
  final SharedPreferences pref;

  const PrefStateManager(this.pref);

  @override
  bool shouldShow(String groupId, TipItem item) {
    final lastVersion = pref.getInt('$groupId.${item.id}');
    return lastVersion == null ? true : lastVersion < item.version;
  }

  @override
  Future<void> tipRead(String groupId, TipItem item) {
    return pref.setInt('$groupId.${item.id}', item.version);
  }
}
```

Example of using [Hive](https://pub.dev/packages/hive)

```dart
void initialize() async {
  final box = Hive.box('myBox');
  OrderedTip.stateManager = HiveStateManager(box);
}

class HiveStateManager extends StateManager {
  final Box box;

  const HiveStateManager(this.box);

  @override
  bool shouldShow(String groupId, TipItem item) {
    final lastVersion = box.get('$groupId.${item.id}');
    return lastVersion == null ? true : lastVersion < item.version;
  }

  @override
  Future<void> tipRead(String groupId, TipItem item) {
    return box.put('$groupId.${item.id}', version);
  }
}
```

## How it works

Main idea is inspired from [`Tooltip`](https://api.flutter.dev/flutter/material/Tooltip-class.html), which using [`overlayState.insert`](https://api.flutter.dev/flutter/widgets/OverlayState/insert.html) to show tips.

`SimpleTip` build tips and _backdrop_ on the `Overlay`, and detect click on backdrop to close it. Multiple tips should build single backdrop and close all tips once it been clicked.

You can also close it by

- User tap `closeButton`
- `dispose` or `deactivate` has been fired

## LICENSE

See in [LICENSE](LICENSE)


CHANGELOG.md
## v0.3.2 2021-12-19

- Allow disabled tips in grouper
- Allow null grouper

## v0.3.1 2021-11-27

- Allow disabled tips in grouper
- Allow null grouper

## v0.3.0 2021-11-27

- Add TipGrouper

## v0.2.1 2021-08-25

- Fix pubspec repository link

## v0.2.0 2021-08-25

- change StateManager interface
- Fix bugs

## v0.1.0 2021-08-24

- Use StateManager instead of static methods

## 0.1.0-dev.1 2021-08-23

- First reviewed release


lib/src/simple_tip.dart
import 'dart:async';

import 'package:flutter/material.dart';

import 'tip_content.dart';

/// A widget for showing tips to it's child and automatically setup position.
///
/// The idea is mainly copy from material [Tooltip].
///
/// Example:
/// ```
/// Widget build(context) {
///   return Tip(
///     message: 'Now it's time to see your profile!',
///     child: TextButton(
///       onPressed: navigateToProfile,
///       child: Text('See profile'),
///     ),
///   );
/// }
/// ```
class SimpleTip extends StatefulWidget {
  /// The title to display in the tip.
  final String? title;

  /// The text to display in the tip.
  ///
  /// This value will also use for semantics.
  final String? message;

  /// Content of tip
  ///
  /// This will "win" when [message] and [content] both set.
  ///
  /// Example:
  /// ```
  /// final content = Text('hi');
  /// ```
  final Widget? content;

  /// Tip's content constraints
  ///
  /// Default:
  /// ```
  /// BoxConstraints(minHeight: 24.0)
  /// ```
  final BoxConstraints boxConstraints;

  /// Tip's container decoration
  ///
  /// Default:
  /// ```
  /// ShapeDecoration(
  ///   color: (isDark ? Colors.white : Colors.grey[700]!).withOpacity(0.9),
  ///   shape: TipShapeBorder(arrowArc: 0.1, target: target),
  /// )
  /// ```
  final Decoration? decoration;

  /// Tip's text default style
  ///
  /// Default:
  /// ```
  /// TextStyle()
  /// ```
  final TextStyle textStyle;

  /// A callback after the start of closing tip.
  ///
  /// This will be needed if you handle tip by filesystem, eg: SharedPreferences, hive
  ///
  /// Example:
  /// ```
  /// final isDisabled = instance.read('mytip') == 1;
  /// final onClosed = () {
  ///   instance.write('mytip', 1);
  /// }
  /// ```
  final VoidCallback onClosed;

  /// The amount of space by which to inset the tip's content.
  ///
  /// Default:
  /// ```
  /// EdgeInsets.all(8.0)
  /// ```
  final EdgeInsets padding;

  /// The empty space that surrounds the tip.
  ///
  /// Defines the tip's outer [Container.margin]. By default, a
  /// long tip will span the width of its window. If long enough,
  /// a tip might also span the window's height. This property allows
  /// one to define how much space the tip must be inset from the edges
  /// of their display window.
  ///
  /// Default:
  /// ```
  /// EdgeInsets.symmetric(horizontal: 16.0)
  /// ```
  final EdgeInsets margin;

  /// The vertical gap between the widget and the displayed tip.
  ///
  /// When [preferBelow] is set to true and tips have sufficient space to
  /// display themselves, this property defines how much vertical space
  /// tips will position themselves under their corresponding widgets.
  /// Otherwise, tips will position themselves above their corresponding
  /// widgets with the given offset.
  ///
  /// Default: `24.0`
  final double verticalOffset;

  /// Text of button to close tip.
  ///
  /// Default: `OK`
  final String closerText;

  /// The length of time that a tip will wait for showing.
  ///
  /// Defaults to 300 milliseconds (same as PageRoute default `transitionDuration`).
  final Duration waitDuration;

  /// Whether the tip's [message] should be excluded from the semantics
  /// tree.
  ///
  /// Defaults to false. A tip will add a [Semantics] label that is set to
  /// [SimpleTip.message]. Set this property to true if the app is going to
  /// provide its own custom semantics label.
  final bool excludeFromSemantics;

  /// Disable tip.
  ///
  /// Wrap [SimpleTip] with `StatefulWidget` and dynamically set this value.
  ///
  /// Default: `false`
  final bool isDisabled;

  /// Whether the tips defaults to being displayed below the widget.
  ///
  /// Defaults to true. If there is insufficient space to display the tip in
  /// the preferred direction, the tip will be displayed in the opposite
  /// direction.
  final bool preferBelow;

  /// Whether use backdrop
  ///
  /// Default: `false`
  final bool withBackdrop;

  /// The widget below this widget in the tree.
  ///
  /// {@macro flutter.widgets.ProxyWidget.child}
  final Widget child;

  const SimpleTip({
    Key? key,
    this.title,
    this.message,
    this.content,
    this.boxConstraints = const BoxConstraints(minHeight: 24.0),
    this.decoration,
    this.textStyle = const TextStyle(),
    required this.onClosed,
    this.padding = const EdgeInsets.all(8.0),
    this.margin = const EdgeInsets.symmetric(horizontal: 16.0),
    this.verticalOffset = 24.0,
    this.closerText = 'OK',
    this.waitDuration = const Duration(milliseconds: 300),
    this.excludeFromSemantics = false,
    this.isDisabled = false,
    this.preferBelow = true,
    this.withBackdrop = false,
    required this.child,
  })  : assert(message != null || content != null),
        super(key: key);

  @override
  SimpleTipState createState() => SimpleTipState();
}

/// State o [SimpleTip]
class SimpleTipState extends State<SimpleTip>
    with SingleTickerProviderStateMixin {
  late AnimationController _controller;
  // save in state avoid missing widget when removing tips
  late bool _withBackdrop;
  OverlayEntry? _entry;
  Timer? _showTimer;

  static OverlayEntry? _backdrop;
  static bool _backdropIsBuilt = false;
  static final List<VoidCallback> _backdropCb = <VoidCallback>[];

  @override
  Widget build(BuildContext context) {
    if (widget.isDisabled) {
      _removeEntry();
      return widget.child;
    }
    assert(Overlay.of(context, debugRequiredFor: widget) != null);
    _showTimer ??= Timer(widget.waitDuration, showEntry);

    return Semantics(
      label: widget.excludeFromSemantics ? null : widget.message,
      child: widget.child,
    );
  }

  @override
  void deactivate() {
    _removeEntry();
    _removeBackdrop();
    _showTimer?.cancel();
    super.deactivate();
  }

  @override
  void dispose() {
    _removeEntry();
    _removeBackdrop();
    _controller.dispose();
    super.dispose();
  }

  /// Gracefully closing tip
  ///
  /// It will trigger [AnimationController.reverse]
  void hideEntry() {
    _controller.reverse();
  }

  @override
  void initState() {
    super.initState();
    _withBackdrop = widget.withBackdrop;
    _controller = AnimationController(
      duration: const Duration(milliseconds: 150),
      reverseDuration: const Duration(milliseconds: 75),
      vsync: this,
    )..addStatusListener(_handleStatusChanged);
  }

  /// Shows the tip
  void showEntry() {
    _showTimer?.cancel();
    _showTimer = null;
    _createNewEntry();
    _controller.forward();
  }

  void _createNewEntry() {
    final overlayState = Overlay.of(context)!;

    final box = context.findRenderObject()! as RenderBox;
    final target = box.localToGlobal(
      box.size.center(Offset.zero),
      ancestor: overlayState.context.findRenderObject(),
    );

    // insert backdrop first
    if (_withBackdrop) {
      if (!_backdropIsBuilt) {
        // We create this widget outside of the overlay entry's builder to prevent
        // updated values from happening to leak into the overlay when the overlay
        // rebuilds.
        final Widget backdropOverlay = SizedBox.expand(
          child: GestureDetector(onTap: () {
            // avoid ConcurrentModificationError: changing iteration during iterate
            for (var cb in _backdropCb) {
              cb();
            }
            _removeBackdrop();
          }),
        );
        _backdrop = OverlayEntry(builder: (_) => backdropOverlay);
        overlayState.insert(_backdrop!);
        _backdropIsBuilt = true;
        _backdropCb.add(hideEntry);
      } else {
        _backdropCb.add(hideEntry);
      }
    }

    final Widget overlay = Directionality(
      textDirection: Directionality.of(context),
      child: TipContent(
        // user provide
        title: widget.title,
        message: widget.message,
        preferBelow: widget.preferBelow,
        boxConstraints: widget.boxConstraints,
        closerText: widget.closerText,
        content: widget.content,
        decoration: widget.decoration,
        textStyle: widget.textStyle,
        margin: widget.margin,
        padding: widget.padding,
        verticalOffset: widget.verticalOffset,
        // package provide
        animation: CurvedAnimation(
          parent: _controller,
          curve: Curves.fastOutSlowIn,
        ),
        target: target,
        hideTip: hideEntry,
      ),
    );
    _entry = OverlayEntry(builder: (BuildContext context) => overlay);
    overlayState.insert(_entry!);
  }

  void _handleStatusChanged(AnimationStatus status) {
    if (status == AnimationStatus.dismissed) {
      _removeEntry();
      widget.onClosed();
    }
  }

  void _removeEntry() {
    if (_entry != null) {
      _showTimer?.cancel();
      _showTimer = null;
      _entry!.remove();
      _entry = null;
    }
  }

  void _removeBackdrop() {
    _backdrop?.remove();
    _backdrop = null;
    _backdropIsBuilt = false;
    _backdropCb.clear();
  }
}
